### PR TITLE
HPCC-31004 Add some stack trace information to exceptions passed back from Python

### PR DESCRIPTION
**Pull request by richardkchapman**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by richardkchapman** [comment]
This may help track down the origin of the strange error being seen in HPCC-31004 - but it may not be generally desirable. I can't as yet work out how to set it via an option on the EMBED statement (would require a bit more refactoring, I fear).

File: plugins/py3embed/py3embed.cpp [hunk 1]
```diff
@@ -44,6 +44,11 @@
 #include "enginecontext.hpp"
 #include <regex>
 
+#if defined (__linux__) || defined(__FreeBSD__)  || defined(__APPLE__)
+#include <execinfo.h> // comment out if not present
```
Comment (ghalliday) [inline plugins/py3embed/py3embed.cpp:48]: info: Gordon pointed me to a new c++17 feature:

#if __has_include(<span>)

To check if an include file exists - that might be useful here.

File: plugins/py3embed/py3embed.cpp [hunk 2]
```diff
@@ -131,6 +136,14 @@ static void failx(const char *message, ...)
     StringBuffer msg;
     msg.append("pyembed: ").valist_appendf(message,args);
     va_end(args);
+#ifdef HAS_BACKTRACE
+    void *stack[5];
+    unsigned nFrames = backtrace(stack, 5);
+    char** strs = backtrace_symbols(stack, nFrames);
+    for (unsigned i = 0; i < nFrames; ++i)
+        msg.append("\n  ").append(strs[i]);
```
Comment (ghalliday) [inline plugins/py3embed/py3embed.cpp:144]: Will the newlines mess up the logging?  Would it be better to have them all on the same line?

**Review by ghalliday — COMMENTED** [review]
Looks like a reasonable change.  One question about newlines.

File: plugins/py3embed/py3embed.cpp (hunk 2 above)
Comment (richardkchapman) [reply plugins/py3embed/py3embed.cpp:144]: It's very ugly on one line. This string is not logged unless the ECL code outputs it, AFAIK

**Review by mckellyln — APPROVED** [review]
I am ok with this.

**Comment by GordonSmith** [comment]
@richardkchapman - FYI - https://track.hpccsystems.com/browse/HPCC-31112

**Pull request merged**